libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshControllerTest.kt

```diff
@@ -0,0 +1,645 @@
+        val route2 = mockk<NavigationRoute>(relaxed = true)
+        val routes = listOf(route1, route2)
+        every {
+            RouteRefreshValidator.validateRoute(any())
+        } returns RouteRefreshValidator.RouteValidationResult.Valid
+
+        sut.startRoutesRefreshing(routes)
+        finishRequest(
+            RouteRefresherResult(
+                false,
+                listOf(route1, route2),
+                RouteProgressData(1, 2, 3)
+            )
+        )
+
+        verify(exactly = 1) { retryStrategy.onNextAttempt() }
+    }
+
+    @Test
+    fun finishRequestSuccessfully() {
+        val route1 = mockk<NavigationRoute>(relaxed = true)
+        val route2 = mockk<NavigationRoute>(relaxed = true)
+        val routes = listOf(route1, route2)
+        every {
+            RouteRefreshValidator.validateRoute(any())
+        } returns RouteRefreshValidator.RouteValidationResult.Valid
+
+        sut.startRoutesRefreshing(routes)
+        val result = RouteRefresherResult(true, listOf(route1, route2), RouteProgressData(1, 2, 3))
+        finishRequest(result)
+
+        verify(exactly = 1) {
+            stateHolder.onSuccess()
+            listener.onRoutesRefreshed(result)
+        }
+        verify { cancellableHandler wasNot Called }
+    }
+
+    @Test
+    fun finishRequestUnsuccessfullyShouldRetry() {
+        val route1 = mockk<NavigationRoute>(relaxed = true)
+        val route2 = mockk<NavigationRoute>(relaxed = true)
+        val routes = listOf(route1, route2)
+        every {
+            RouteRefreshValidator.validateRoute(any())
+        } returns RouteRefreshValidator.RouteValidationResult.Valid
+        every { retryStrategy.shouldRetry() } returns true
+
+        sut.startRoutesRefreshing(routes)
+        clearMocks(retryStrategy, answers = false)
+        val result = RouteRefresherResult(false, listOf(route1, route2), RouteProgressData(1, 2, 3))
+        finishRequest(result)
+
+        verify(exactly = 1) {
+            cancellableHandler.postDelayed(interval, any(), any())
+        }
+        verify(exactly = 0) {
+            stateHolder.onFailure(any())
+            listener.onRoutesRefreshed(any())
+            retryStrategy.reset()
+        }
+    }
+
+    @Test
+    fun finishRequestUnsuccessfullyShouldRetryDoesNotNotifyOnStart() {
+        val route1 = mockk<NavigationRoute>(relaxed = true)
+        val route2 = mockk<NavigationRoute>(relaxed = true)
+        val routes = listOf(route1, route2)
+        every {
+            RouteRefreshValidator.validateRoute(any())
+        } returns RouteRefreshValidator.RouteValidationResult.Valid
+        every { retryStrategy.shouldRetry() } returns true
+
+        sut.startRoutesRefreshing(routes)
+        val result = RouteRefresherResult(false, listOf(route1, route2), RouteProgressData(1, 2, 3))
+        finishRequest(result)
+        startRequest()
+
+        verify(exactly = 0) { stateHolder.onStarted() }
```

Comment:
I understand the retry shouldn't call start again but here it looks like we didn't call start at all.

Reply:
That's because I did `finishRequest` without `startRequest`. First `startRequest` is not required for the unit test.